### PR TITLE
Add `typespec/1` macro to transform module behaviour

### LINKED ISSUE #382: Types using protobuf transformer have wrong typespecs
When a protobuf module implements `transform_module/0`, the autogenerated typespecs usually become invalid. 

Assume the following example (extracted from `Protobuf.TransformModule` docs):

```elixir
defmodule StringMessage do
  use Protobuf, syntax: :proto3

  field :value, 1, type: :string

  def transform_module(), do: MyTransformModule
end
```

With the following transformer (also extracted):
```elixir
defmodule MyTransformModule do
  @behaviour Protobuf.TransformModule

  @impl true
  def encode(string, StringMessage) when is_binary(string), do: %StringMessage{value: string}

  @impl true
  def decode(%StringMessage{value: string}, StringMessage), do: string
end
```

You can use the above type as a string:
```
iex(4)> defmodule Foo do
...(4)>     use Protobuf, syntax: :proto3
...(4)>     field :string, 1, type: StringMessage
...(4)> end
iex(5)> Protobuf.encode(%Foo{string: "hello"})
<<10, 7, 10, 5, 104, 101, 108, 108, 111>>
iex(6)> Protobuf.decode(v, Foo)
%Foo{string: "hello", __unknown_fields__: []}
```

If you try to use a struct, it fails:
```
iex(7)> Protobuf.encode(%Foo{string: %StringMessage{value: "hello"}})
** (Protobuf.EncodeError) Got error when encoding Foo#string: ** (FunctionClauseError) no function clause matching in MyTransformModule.encode/2
```

Yet the auto-generated typespec doesn't reflect that. Instead, the typespec will be:
```elixir
@type t() :: %__MODULE__{value: String.t()}
```

This causes valid code to be flagged by dialyzer as invalid, and invalid code to be considered valid.

### PR DESCRIPTION
**Pull request by v0idpwn**
Adds a `typespec/1` macrocallback to transform module behaviour.

This macro can modify the typespec for modules using it. This way, we can provide correct typespecs for modules using transformers. A "breaking" change is that transform modules can't depend on Protobuf definitions anymore. This is quite easily solvable, though.

Closes #382 

--------------------------------------------------
**Previous description, from previous implementation:**

We detect that `transform_module/0` was defined through a on_definition hook. We then store the module in an attribute. This attribute is later used by the `before_compile` callback to set the typespec, which is set in the form of:

```elixir
@type t() :: transform_module.t(__MODULE__)
```

Then, transform modules can implement proper typespecs.

No changes if `transform_module` is not overriden.

Closes #382

### CODE REVIEW

**Comment by v0idpwn** [comment]
@whatyouhide @ericmj if you agree on approach here, I'll add a few tests, documentation, etc :)

File: lib/protobuf/dsl.ex [hunk 1]
```diff
@@ -226,19 +242,25 @@ defmodule Protobuf.DSL do
     end
   end
 
-  defp def_t_typespec(%MessageProps{enum?: true} = props, _extension_props) do
+  defp def_t_typespec(_props, _extension_props, transform_module_ast) when not is_nil(transform_module_ast) do
+    quote do
+      @type t() :: unquote(transform_module_ast).t(__MODULE__)
```
Comment (v0idpwn) [inline lib/protobuf/dsl.ex:247]: The biggest potential issue is that we are assuming that the function ast will always be a module. But I can't see it being something else if you are using the library properly :)

**Comment by v0idpwn** [comment]
I tested this in our codebases and it resolves dialyzer issues. We are using transformers for, for example, well known types.

**Comment by v0idpwn** [comment]
JFYI, I don't think the typespecs as they stand in this PR are good enough. I'm working on a new implementation that, although more complex, provides better typespecs.

File: lib/protobuf/dsl.ex [hunk 2]
```diff
@@ -226,19 +242,50 @@ defmodule Protobuf.DSL do
...
+       when not is_nil(transform_module_ast) do
+    default_typespec = def_t_typespec(props, extension_props, nil)
+
+    quote do
+      case Code.ensure_compiled(unquote(transform_module_ast)) do
+        {:module, _} ->
+          :ok
+
+        _ ->
+          raise CompileError, """
```
Comment (v0idpwn) [inline lib/protobuf/dsl.ex:255]: Need to add a test for this error

File: lib/protobuf/transform_module/infer_fields_from_enum.ex [hunk 3]
```diff
@@ -5,6 +5,8 @@ defmodule Protobuf.TransformModule.InferFieldsFromEnum do
   `c:Protobuf.new/1` before the value is encoded.
   """
 
+  @type t(_module) :: term()
+
```
Comment (v0idpwn) [inline lib/protobuf/transform_module/infer_fields_from_enum.ex:9]:
```suggestion
```

File: lib/protobuf/dsl.ex [hunk 4]
```diff
@@ -226,19 +242,50 @@ defmodule Protobuf.DSL do
...
+          Transform module #{inspect(unquote(transform_module_ast))} not available
+          during protobuf definition compilation.
```
[the remark this answers is not in the document]
Comment (whatyouhide) [reply lib/protobuf/dsl.ex:257]: ...and all instances below

File: lib/protobuf/transform_module.ex [hunk 5]
```diff
@@ -50,4 +60,13 @@ defmodule Protobuf.TransformModule do
   Called after a message is decoded.
   """
   @callback decode(message(), type()) :: value()
+
+  @doc """
+  Transforms the typespec for modules using this transformer
```
Comment (whatyouhide) [inline lib/protobuf/transform_module.ex:65]:
```suggestion
  Transforms the typespec for modules using this transformer.
```

File: lib/protobuf/transform_module.ex [hunk 6]
```diff
@@ -50,4 +60,13 @@ defmodule Protobuf.TransformModule do
...
+
+  Optional. If this callback is not present, the default typespec will be used.
```
Comment (whatyouhide) [inline lib/protobuf/transform_module.ex:67]: ExDoc already marks this as optional thanks to `@optional_callbacks`

```suggestion
  If this callback is not present, the default typespec will be used.
```

**Comment by v0idpwn** [comment]
@whatyouhide I believe I addressed all your comments, plus I made a few simplifications. I'll go ahead and merge this :)

**Pull request merged**